model/jpa/src/main/java/org/keycloak/models/jpa/JpaUserProvider.java

```diff
@@ -700,54 +702,80 @@ public int getUsersCount(RealmModel realm) {
+
         for (Map.Entry<String, String> entry : attributes.entrySet()) {
-            String parameterName = null;
-            if (entry.getKey().equals(UserModel.USERNAME)) {
-                parameterName = JpaUserProvider.USERNAME;
-            } else if (entry.getKey().equalsIgnoreCase(UserModel.FIRST_NAME)) {
-                parameterName = JpaUserProvider.FIRST_NAME;
-            } else if (entry.getKey().equalsIgnoreCase(UserModel.LAST_NAME)) {
-                parameterName = JpaUserProvider.LAST_NAME;
-            } else if (entry.getKey().equalsIgnoreCase(UserModel.EMAIL)) {
-                parameterName = JpaUserProvider.EMAIL;
+            String key = entry.getKey();
+            String value = entry.getValue();
+
+            if (value == null) {
+                continue;
+            }
+
+            value = value.toLowerCase();
+
+            if (key.equals(UserModel.USERNAME)) {
+                predicates.add(builder.like(builder.lower(root.get(UserModel.USERNAME)), "%" + value + "%"));
+            } else if (key.equalsIgnoreCase(UserModel.FIRST_NAME)) {
+                predicates.add(builder.like(builder.lower(root.get(UserModel.FIRST_NAME)), "%" + value + "%"));
+            } else if (key.equalsIgnoreCase(UserModel.LAST_NAME)) {
+                predicates.add(builder.like(builder.lower(root.get(UserModel.LAST_NAME)), "%" + value + "%"));
+            } else if (key.equalsIgnoreCase(UserModel.EMAIL)) {
+                predicates.add(builder.like(builder.lower(root.get(UserModel.EMAIL)), "%" + value + "%"));
+            } else if (UserModel.GROUPS.equals(key)) {
+                Subquery<String> subquery = queryBuilder.subquery(String.class);
+                Root<UserGroupMembershipEntity> from = subquery.from(UserGroupMembershipEntity.class);
+                Join<Object, Object> user1 = from.join("user");
+                Join<Object, Object> user = user1.on(builder.equal(user1.get("id"), root.get("id")));
+                subquery.select(user.get("id"));
+                List<Predicate> subPredicates = new ArrayList<>();
+
+                try {
+                    List list = JsonSerialization.readValue(value, List.class);
+
+                    for (Object o : list) {
+                        subPredicates.add(builder.equal(from.get("groupId"), o));
+                    }
+                } catch (IOException e) {
+                    e.printStackTrace();
+                }
+
+                subquery.where(subPredicates.toArray(new Predicate[subPredicates.size()]));
+
+                predicates.add(root.get("id").in(subquery));
             }
-            if (parameterName == null) continue;
-            query.setParameter(parameterName, "%" + entry.getValue().toLowerCase() + "%");
         }
+
+        queryBuilder.where(predicates.toArray(new Predicate[predicates.size()])).orderBy(builder.asc(root.get(UserModel.USERNAME)));
+
+        Query query = em.createQuery(queryBuilder);
```

Comment:
I find the whole query generation code extremely hard to read and understand. Wouldn't it be possible to at least add a comment sketching the generated query? I do prefer readable code over comments any day but this is too much for me or I did not have enough coffee, yet...